pkg/util/ec2/imds_helpers.go

```diff
@@ -57,39 +66,50 @@ func getMetadataItemWithMaxLength(ctx context.Context, endpoint string, forceIMD
 	return result, err
 }
 
-func getMetadataItem(ctx context.Context, endpoint string, forceIMDSv2 bool) (string, error) {
+func getMetadataItem(ctx context.Context, endpoint string, usage IMDSv2Usage) (string, error) {
 	if !pkgconfigsetup.IsCloudProviderEnabled(CloudProviderName, pkgconfigsetup.Datadog()) {
 		return "", fmt.Errorf("cloud provider is disabled by configuration")
 	}
 
-	return doHTTPRequest(ctx, metadataURL+endpoint, forceIMDSv2)
+	return doHTTPRequest(ctx, metadataURL+endpoint, usage)
 }
 
 // UseIMDSv2 returns true if the agent should use IMDSv2
-func UseIMDSv2(forceIMDSv2 bool) bool {
-	return pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || forceIMDSv2
+func UseIMDSv2(force bool, disable bool) IMDSv2Usage {
+	// if force is set, we use IMDSv2, highest priority
+	if force {
+		return useIMDSv2
+		// if disable is set, we disable IMDSv2
+	} else if disable {
+		return useIMDSv1
+		// if ec2_prefer_imdsv2 is set, we use IMDSv2
+	} else if pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || pkgconfigsetup.Datadog().GetBool("ec2_imdsv2_transition") {
+		return useIMDSv2Andv1
+	}
+	// if nothing indicates to use IMDSv2, we default to IMDSv1
+	return useIMDSv1
 }
 
-func doHTTPRequest(ctx context.Context, url string, forceIMDSv2 bool) (string, error) {
+func doHTTPRequest(ctx context.Context, url string, imdsv2Usage IMDSv2Usage) (string, error) {
 	source := metadataSourceIMDSv1
 	headers := map[string]string{}
-	if UseIMDSv2(forceIMDSv2) {
+	if imdsv2Usage == useIMDSv2Andv1 || imdsv2Usage == useIMDSv2 {
```

Comment:
I think this logic may need a bit of clarification (maybe just naming to clarify things). As it stands, it's very unclear what `imdsv2Usage` stands for without scrolling up to see what the enum contains. Maybe something like:
```go
if allowedIMDSVersions == IMDSAllVersions || allowedIMDSVersions == IMDSVersion2 {
```

Alternatively, maybe this is something that can be improved with a real object but this may be too much to do on short notice:
```go
func doHTTPRequest(ctx context.Context, url string, ec2IMDSVersionConfig EC2IMDSVersionConfig) (string, error) {
...
if ec2IMDSVersionConfig.enablesV2() {
...
  if ec2IMDSVersionConfig.forcesV2() {
  ...
```

Reply:
renaming done with some small changes

Reply:
I will do something like that:

```
func doHTTPRequest(ctx context.Context, url string, ec2IMDSVersionConfig EC2IMDSVersionConfig) (string, error) {
...
if ec2IMDSVersionConfig.enablesV2() {
...
  if ec2IMDSVersionConfig.forcesV2() {
  ...
```

As you mentioned in a future PR (after the test PR being done) since it may brings some behaviour changes if it works for you ?

Reply:
done in: [`9fdf71d` (#30760)](https://github.com/DataDog/datadog-agent/pull/30760/commits/9fdf71d23d34c820a57b04c02fa20e4c74a558f5)